File: Day47/Part5/hex_arena.py
import heapq

def solve_hex_arena(target, blocked, runes):
    tx, ty, tz = target
    blocked_set = set(tuple(b) for b in blocked)
    rune_set = set(tuple(r) for r in runes)
    runes_list = [tuple(r) for r in runes]
    
    # Directions in Cube Coordinates
    # x+y+z = 0
    directions = [
        (1, -1, 0), (1, 0, -1), (0, 1, -1),
        (-1, 1, 0), (-1, 0, 1), (0, -1, 1)
    ]
    
    pq = [(0, 0, 0, 0)] # cost, x, y, z
    dist = {}
    dist[(0, 0, 0)] = 0
    
    limit = 50
    
    while pq:
        cost, x, y, z = heapq.heappop(pq)
        
        if x == tx and y == ty and z == tz:
            return cost
        
        if cost > dist.get((x, y, z), float('inf')):
            continue
        
        # 1. Walk (Cost 1)
        for dx, dy, dz in directions:
            nx, ny, nz = x + dx, y + dy, z + dz
            if max(abs(nx), abs(ny), abs(nz)) <= limit and (nx, ny, nz) not in blocked_set:
                new_cost = cost + 1
                if new_cost < dist.get((nx, ny, nz), float('inf')):
                    dist[(nx, ny, nz)] = new_cost
                    heapq.heappush(pq, (new_cost, nx, ny, nz))
                    
        # 2. Blink (Cost 2)
        for dx, dy, dz in directions:
            nx, ny, nz = x + dx, y + dy, z + dz
            while max(abs(nx), abs(ny), abs(nz)) <= limit and (nx, ny, nz) not in blocked_set:
                new_cost = cost + 2
                if new_cost < dist.get((nx, ny, nz), float('inf')):
                    dist[(nx, ny, nz)] = new_cost
                    heapq.heappush(pq, (new_cost, nx, ny, nz))
                nx += dx
                ny += dy
                nz += dz
                
        # 3. Rune Teleport (Cost 1)
        if (x, y, z) in rune_set:
            for rx, ry, rz in runes_list:
                if (rx, ry, rz) == (x, y, z):
                    continue
                new_cost = cost + 1
                if new_cost < dist.get((rx, ry, rz), float('inf')):
                    dist[(rx, ry, rz)] = new_cost
                    heapq.heappush(pq, (new_cost, rx, ry, rz))
                    
    return -1

if __name__ == "__main__":
    Target = [0, -2, 2]
    Blocked = [[0, -1, 1]]
    Runes = [[1, -1, 0], [0, -2, 2]]
    
    result = solve_hex_arena(Target, Blocked, Runes)
    print(f"Minimum Cost: {result}")

File: learnLeetCode/Day47/Part5/hex_arena.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <set>
#include <cmath>
#include <map>

using namespace std;

struct Point {
    int x, y, z;

    bool operator<(const Point& other) const {
        if (x != other.x) return x < other.x;
        if (y != other.y) return y < other.y;
        return z < other.z;
    }

    bool operator==(const Point& other) const {
        return x == other.x && y == other.y && z == other.z;
    }
};

struct State {
    int cost;
    Point p;

    bool operator>(const State& other) const {
        return cost > other.cost;
    }
};

int solveHexArena(Point target, vector<Point>& blocked, vector<Point>& runes) {
    set<Point> blocked_set(blocked.begin(), blocked.end());
    set<Point> rune_set(runes.begin(), runes.end());

    priority_queue<State, vector<State>, greater<State>> pq;
    pq.push({0, {0, 0, 0}});

    map<Point, int> dist;
    dist[{0, 0, 0}] = 0;

    int limit = 50;

    int dx[] = {1, 1, 0, -1, -1, 0};
    int dy[] = {-1, 0, 1, 1, 0, -1};
    int dz[] = {0, -1, -1, 0, 1, 1};

    while (!pq.empty()) {
        State current = pq.top();
        pq.pop();

        int cost = current.cost;
        Point p = current.p;

        if (p == target) {
            return cost;
        }

        if (dist.count(p) && cost > dist[p]) {
            continue;
        }

        // 1. Walk
        for (int i = 0; i < 6; ++i) {
            Point next = {p.x + dx[i], p.y + dy[i], p.z + dz[i]};
            if (max(abs(next.x), max(abs(next.y), abs(next.z))) <= limit && blocked_set.find(next) == blocked_set.end()) {
                int new_cost = cost + 1;
                if (!dist.count(next) || new_cost < dist[next]) {
                    dist[next] = new_cost;
                    pq.push({new_cost, next});
                }
            }
        }

        // 2. Blink
        for (int i = 0; i < 6; ++i) {
            Point next = {p.x + dx[i], p.y + dy[i], p.z + dz[i]};
            while (max(abs(next.x), max(abs(next.y), abs(next.z))) <= limit && blocked_set.find(next) == blocked_set.end()) {
                int new_cost = cost + 2;
                if (!dist.count(next) || new_cost < dist[next]) {
                    dist[next] = new_cost;
                    pq.push({new_cost, next});
                }
                next.x += dx[i];
                next.y += dy[i];
                next.z += dz[i];
            }
        }

        // 3. Rune Teleport
        if (rune_set.count(p)) {
            for (const auto& r : runes) {
                if (r == p) continue;
                int new_cost = cost + 1;
                if (!dist.count(r) || new_cost < dist[r]) {
                    dist[r] = new_cost;
                    pq.push({new_cost, r});
                }
            }
        }
    }

    return -1;
}

int main() {
    Point Target = {0, -2, 2};
    vector<Point> Blocked = {{0, -1, 1}};
    vector<Point> Runes = {{1, -1, 0}, {0, -2, 2}};

    int result = solveHexArena(Target, Blocked, Runes);
    cout << "Minimum Cost: " << result << endl;

    return 0;
}
